File: legacy-csharp-avalonia/src/Models/History.cs
namespace PixelSpriteEditor.Models;

/// <summary>
/// Stroke-based undo/redo. A "stroke" is one user gesture (a pencil drag
/// from mouse-down to mouse-up, a single fill click, etc). Each stroke
/// records only the pixels it actually touched so undo/redo stays cheap
/// even on large canvases.
/// </summary>
public sealed class History
{
    public readonly record struct PixelChange(int X, int Y, PixelColor Before, PixelColor After);

    private List<PixelChange> _inProgress = new();
    private readonly List<List<PixelChange>> _undoStack = new();
    private readonly List<List<PixelChange>> _redoStack = new();

    public void BeginStroke() => _inProgress = new List<PixelChange>();

    /// <summary>Call after the pixel has already been written to the canvas.</summary>
    public void RecordChange(int x, int y, PixelColor before, PixelColor after) =>
        _inProgress.Add(new PixelChange(x, y, before, after));

    public void EndStroke()
    {
        if (_inProgress.Count == 0) return;
        _undoStack.Add(_inProgress);
        _inProgress = new List<PixelChange>();
        _redoStack.Clear();
    }

    public bool CanUndo => _undoStack.Count > 0;
    public bool CanRedo => _redoStack.Count > 0;

    public void Undo(PixelCanvas canvas)
    {
        if (_undoStack.Count == 0) return;
        var stroke = _undoStack[^1];
        _undoStack.RemoveAt(_undoStack.Count - 1);
        for (int i = stroke.Count - 1; i >= 0; i--) canvas.SetPixel(stroke[i].X, stroke[i].Y, stroke[i].Before);
        _redoStack.Add(stroke);
    }

    public void Redo(PixelCanvas canvas)
    {
        if (_redoStack.Count == 0) return;
        var stroke = _redoStack[^1];
        _redoStack.RemoveAt(_redoStack.Count - 1);
        foreach (var change in stroke) canvas.SetPixel(change.X, change.Y, change.After);
        _undoStack.Add(stroke);
    }

    public void Clear()
    {
        _inProgress.Clear();
        _undoStack.Clear();
        _redoStack.Clear();
    }
}

File: PixelSpriteEditor/src/History.java
import java.util.ArrayList;
import java.util.List;

/**
 * Stroke-based undo/redo. A "stroke" is one user gesture (a pencil drag from
 * mouse-down to mouse-up, a single fill click, a committed line, etc). Each
 * stroke records only the pixels it actually touched so undo/redo stays
 * cheap even on larger canvases.
 */
public class History {

    /** One pixel's before/after ARGB value within a stroke. */
    public static final class PixelChange {
        final int x, y, before, after;

        PixelChange(int x, int y, int before, int after) {
            this.x = x;
            this.y = y;
            this.before = before;
            this.after = after;
        }
    }

    private List<PixelChange> inProgress = new ArrayList<>();
    private final List<List<PixelChange>> undoStack = new ArrayList<>();
    private final List<List<PixelChange>> redoStack = new ArrayList<>();

    public void beginStroke() {
        inProgress = new ArrayList<>();
    }

    /** Call after the pixel has already been written to the canvas. */
    public void recordChange(int x, int y, int before, int after) {
        inProgress.add(new PixelChange(x, y, before, after));
    }

    public void endStroke() {
        if (inProgress.isEmpty()) return;
        undoStack.add(inProgress);
        inProgress = new ArrayList<>();
        redoStack.clear();
    }

    public boolean canUndo() {
        return !undoStack.isEmpty();
    }

    public boolean canRedo() {
        return !redoStack.isEmpty();
    }

    public void undo(PixelCanvas canvas) {
        if (undoStack.isEmpty()) return;
        List<PixelChange> stroke = undoStack.remove(undoStack.size() - 1);
        for (int i = stroke.size() - 1; i >= 0; i--) {
            PixelChange c = stroke.get(i);
            canvas.setPixel(c.x, c.y, c.before);
        }
        redoStack.add(stroke);
    }

    public void redo(PixelCanvas canvas) {
        if (redoStack.isEmpty()) return;
        List<PixelChange> stroke = redoStack.remove(redoStack.size() - 1);
        for (PixelChange c : stroke) {
            canvas.setPixel(c.x, c.y, c.after);
        }
        undoStack.add(stroke);
    }

    public void clear() {
        inProgress.clear();
        undoStack.clear();
        redoStack.clear();
    }
}
